File: micro_bench/step408_bench.jl
"""
Minimal benchmark reproducing step 408 of tensornetwork_permutation_light_415.

Binary einsum: abcdefghijxyz,cdefghijklmnopqrstuvwxyz->abklmnopqrstuvwxyz
All dims = 2 (binary tensor network).
m=4, k=256, n=8192, batch=8.

Compares:
  1) OMEinsum pairwise contraction (what the full benchmark uses)
  2) Manual: permutedims! + BLAS gemm (to isolate permutation cost)
  3) Just the permutedims! cost alone
"""

using OMEinsum
using LinearAlgebra

# --- Tensor setup ---
# All dims = 2, column-major. A has 13 dims, B has 24 dims.
# After lazy permutation (metadata-only reorder), strides become scattered.
# This is the situation strided-rs faces.

const NDIMS_A = 13
const NDIMS_B = 24
const NDIMS_C = 18

function make_tensors()
    A = rand(Float64, ntuple(_ -> 2, NDIMS_A)...)
    B = rand(Float64, ntuple(_ -> 2, NDIMS_B)...)
    return A, B
end

# The canonical einsum expression
# A indices:   a b c d e f g h i j x y z    (lo=ab, sum=cdefghij, batch=xyz)
# B indices:   c d e f g h i j k l m n o p q r s t u v w x y z  (sum=cdefghij, ro=klmnopqrstuvw, batch=xyz)
# C indices:   a b k l m n o p q r s t u v w x y z  (lo=ab, ro=klmnopqrstuvw, batch=xyz)

# einsum notation (1-indexed labels for OMEinsum)
# A: (1,2,3,4,5,6,7,8,9,10,24,25,26)
# B: (3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26)
# C: (1,2,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26)

const IA = (1,2,3,4,5,6,7,8,9,10,24,25,26)
const IB = (3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26)
const IC = (1,2,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26)

# --- The permutations that strided-rs applies (0-indexed -> 1-indexed) ---
# After lazy permutation of A from previous steps, A's dims are reordered by left_perm.
# left_perm (0-indexed) = [1, 12, 0, 4, 5, 6, 7, 8, 9, 11, 2, 3, 10]
# This means: canonical dim j uses original dim left_perm[j].
# Julia permutedims(A, perm): result[i1,i2,...] = A[i_{perm[1]}, i_{perm[2]}, ...]
# So Julia perm = left_perm .+ 1
const LEFT_PERM = (2, 13, 1, 5, 6, 7, 8, 9, 10, 12, 3, 4, 11)
const RIGHT_PERM = (5, 11, 24, 13, 21, 1, 4, 18, 2, 3, 7, 8, 9, 10, 12, 14, 15, 16, 19, 22, 23, 6, 17, 20)

# --- Benchmark helpers ---
function bench(f, warmup=3, nruns=15)
    for _ in 1:warmup
        f()
    end
    times = Float64[]
    for _ in 1:nruns
        t = @elapsed f()
        push!(times, t)
    end
    sort!(times)
    med = times[div(length(times)+1, 2)]
    q1 = times[max(1, div(length(times)+1, 4))]
    q3 = times[min(length(times), div(3*(length(times)+1), 4))]
    return med, q3 - q1
end

function main()
    A_orig, B_orig = make_tensors()

    println("Step 408 micro-benchmark (all dims=2)")
    println("A: $(size(A_orig)) = $(length(A_orig)) elements")
    println("B: $(size(B_orig)) = $(length(B_orig)) elements")
    println("=" ^ 70)

    # --- 1) OMEinsum pairwise contraction (DynamicEinCode) ---
    code = DynamicEinCode(collect.([IA, IB]), collect(IC))
    f_omeinsum() = einsum(code, (A_orig, B_orig))
    med, iqr = bench(f_omeinsum)
    C_ref = f_omeinsum()
    println("OMEinsum DynamicEinCode:         $(round(med*1000, digits=3)) ms (IQR $(round(iqr*1000, digits=3)) ms)")

    # --- 2) Simulate strided-rs's situation ---
    # strided-rs has lazy-permuted tensors (scattered strides).
    # Equivalent in Julia: create a PermutedDimsArray (a view with reordered dims)
    # then permutedims! to make contiguous, then BLAS GEMM.

    # Create the "lazy permuted" view (like strided-rs after metadata permutation)
    # Note: PermutedDimsArray(A, perm) creates a view where
    # result[i1,...] = A[i_{perm[1]}, ...] -- same semantics as permutedims
    A_lazy = PermutedDimsArray(A_orig, LEFT_PERM)
    B_lazy = PermutedDimsArray(B_orig, RIGHT_PERM)

    # The "eager copy" path: permutedims! to materialize contiguous data
    A_buf = similar(A_lazy)
    B_buf = similar(B_lazy)

    function copy_and_gemm()
        permutedims!(A_buf, A_orig, LEFT_PERM)
        permutedims!(B_buf, B_orig, RIGHT_PERM)
        # Now reshape for GEMM: A_buf is [lo(2), sum(8), batch(3)] = [4, 256, 8]
        # B_buf is [sum(8), ro(13), batch(3)] = [256, 8192, 8]
        m, k, n, nb = 4, 256, 8192, 8
        A_mat = reshape(A_buf, m, k, nb)
        B_mat = reshape(B_buf, k, n, nb)
        C_mat = Array{Float64}(undef, m, n, nb)
        for bi in 1:nb
            mul!(view(C_mat, :, :, bi), view(A_mat, :, :, bi), view(B_mat, :, :, bi))
        end
        return C_mat
    end
    med2, iqr2 = bench(copy_and_gemm)
    println("permutedims! + BLAS gemm:        $(round(med2*1000, digits=3)) ms (IQR $(round(iqr2*1000, digits=3)) ms)")

    # --- 3) Just the permutedims! cost (B only, since it's 16M elements) ---
    function just_perm_B()
        permutedims!(B_buf, B_orig, RIGHT_PERM)
    end
    med3, iqr3 = bench(just_perm_B)
    println("permutedims!(B) only (16M f64):  $(round(med3*1000, digits=3)) ms (IQR $(round(iqr3*1000, digits=3)) ms)")

    # --- 4) Just the permutedims! cost (A only, 8K elements) ---
    function just_perm_A()
        permutedims!(A_buf, A_orig, LEFT_PERM)
    end
    med4, iqr4 = bench(just_perm_A)
    println("permutedims!(A) only (8K f64):   $(round(med4*1000, digits=3)) ms (IQR $(round(iqr4*1000, digits=3)) ms)")

    # --- 5) Just BLAS gemm (with pre-materialized contiguous data) ---
    permutedims!(A_buf, A_orig, LEFT_PERM)
    permutedims!(B_buf, B_orig, RIGHT_PERM)
    m, k, n, nb = 4, 256, 8192, 8
    A_mat = reshape(A_buf, m, k, nb)
    B_mat = reshape(B_buf, k, n, nb)
    C_mat = Array{Float64}(undef, m, n, nb)
    function just_gemm()
        for bi in 1:nb
            mul!(view(C_mat, :, :, bi), view(A_mat, :, :, bi), view(B_mat, :, :, bi))
        end
    end
    med5, iqr5 = bench(just_gemm)
    println("BLAS gemm only (8 batches):      $(round(med5*1000, digits=3)) ms (IQR $(round(iqr5*1000, digits=3)) ms)")

    # --- Verify correctness ---
    C_manual = copy_and_gemm()
    # C_ref has shape (2,2,...,2) 18 dims, C_manual is (4, 8192, 8)
    # Reshape for comparison
    C_ref_r = reshape(C_ref, m, n, nb)
    err = maximum(abs.(C_ref_r .- C_manual))
    println("\nMax error (OMEinsum vs manual): $err")
end

main()

File: micro_bench/step408_bench.rs
//! Minimal benchmark reproducing step 408 of tensornetwork_permutation_light_415.
//!
//! Binary einsum with all dims = 2 (binary tensor network).
//! m=4, k=256, n=8192, batch=8.
//!
//! The key bottleneck: B has 24 binary dims with scattered strides after
//! canonical reorder. `prepare_input_owned` must copy 16M elements to
//! contiguous layout before BLAS GEMM.
//!
//! Build & run:
//!   cargo run --release --no-default-features --features blas --bin step408_bench
//!   OMP_NUM_THREADS=4 cargo run --release --no-default-features --features blas --bin step408_bench

use std::time::Instant;
use strided_view::StridedArray;

fn bench<F: FnMut()>(mut f: F, warmup: usize, nruns: usize) -> (f64, f64) {
    for _ in 0..warmup {
        f();
    }
    let mut times = Vec::with_capacity(nruns);
    for _ in 0..nruns {
        let t = Instant::now();
        f();
        times.push(t.elapsed().as_secs_f64() * 1000.0);
    }
    times.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let med = times[times.len() / 2];
    let q1 = times[times.len() / 4];
    let q3 = times[3 * times.len() / 4];
    (med, q3 - q1)
}

fn main() {
    let warmup = 3;
    let nruns = 15;

    // A: 13 dims of size 2, col-major (contiguous). 8192 elements.
    let a_dims: Vec<usize> = vec![2; 13];
    let a = StridedArray::<f64>::col_major(&a_dims);
    // B: 24 dims of size 2, col-major (contiguous). 16M elements.
    let b_dims: Vec<usize> = vec![2; 24];
    let b = StridedArray::<f64>::col_major(&b_dims);
    // C: 18 dims of size 2, col-major. 262144 elements.
    let c_dims: Vec<usize> = vec![2; 18];

    // Exact einsum labels reconstructed from step 408's plan
    // These labels produce the same scattered-stride pattern as the real benchmark.
    let ia: Vec<char> = "caxydefghizjb".chars().collect();
    let ib: Vec<char> = "hklicxmnopdqfrstyjuzgvwe".chars().collect();
    let ic: Vec<char> = "abklwmnopqrstuvxyz".chars().collect();

    println!("Step 408 micro-benchmark (all dims=2)");
    println!("A: {:?} = {} elements", a.dims(), a.dims().iter().product::<usize>());
    println!("B: {:?} = {} elements", b.dims(), b.dims().iter().product::<usize>());
    println!("C: output {:?} = {} elements", &c_dims, c_dims.iter().product::<usize>());
    println!("Einsum: {},{}->{}", ia.iter().collect::<String>(), ib.iter().collect::<String>(), ic.iter().collect::<String>());
    println!("{}", "=".repeat(70));

    // --- 1) Full einsum2_into_owned (scattered strides, what the real benchmark does) ---
    let (med, iqr) = bench(|| {
        let mut c_arr = StridedArray::<f64>::col_major(&c_dims);
        strided_einsum2::einsum2_into_owned(
            c_arr.view_mut(), a.clone(), b.clone(),
            &ic, &ia, &ib, 1.0, 0.0, false, false,
        ).unwrap();
    }, warmup, nruns);
    println!("einsum2 full (scattered B):    {:.3} ms (IQR {:.3} ms)", med, iqr);

    // --- 2) Isolate the copy cost: permute B to canonical order, then copy ---
    // Reconstruct the right_perm that einsum2 computes internally
    let right_perm: Vec<usize> = vec![
        4, 10, 23, 12, 20, 0, 3, 17, 1, 2, 6, 7, 8, 9, 11, 13, 14, 15, 18, 21, 22, 5, 16, 19,
    ];
    let b_perm = b.permuted(&right_perm).unwrap();
    println!(
        "\nB after canonical reorder: dims={:?} strides={:?}",
        b_perm.dims(), b_perm.strides()
    );

    let (med_copy_b, iqr_copy_b) = bench(|| {
        let mut b_dest = StridedArray::<f64>::col_major(b_perm.dims());
        strided_perm::copy_into(&mut b_dest.view_mut(), &b_perm.view()).unwrap();
    }, warmup, nruns);
    println!("copy_into B (16M, scattered):  {:.3} ms (IQR {:.3} ms)", med_copy_b, iqr_copy_b);

    // --- 3) Copy cost for A ---
    let left_perm: Vec<usize> = vec![1, 12, 0, 4, 5, 6, 7, 8, 9, 11, 2, 3, 10];
    let a_perm = a.permuted(&left_perm).unwrap();

    let (med_copy_a, iqr_copy_a) = bench(|| {
        let mut a_dest = StridedArray::<f64>::col_major(a_perm.dims());
        strided_perm::copy_into(&mut a_dest.view_mut(), &a_perm.view()).unwrap();
    }, warmup, nruns);
    println!("copy_into A (8K, scattered):   {:.3} ms (IQR {:.3} ms)", med_copy_a, iqr_copy_a);

    // --- 4) einsum2 with pre-contiguous data (isolates GEMM cost) ---
    // Make A and B contiguous in canonical order, then use labels matching canonical order
    let mut a_contig = StridedArray::<f64>::col_major(a_perm.dims());
    strided_perm::copy_into(&mut a_contig.view_mut(), &a_perm.view()).unwrap();
    let mut b_contig = StridedArray::<f64>::col_major(b_perm.dims());
    strided_perm::copy_into(&mut b_contig.view_mut(), &b_perm.view()).unwrap();

    // Labels in canonical order after permutation
    let ia_canon: Vec<char> = left_perm.iter().map(|&i| ia[i]).collect();
    let ib_canon: Vec<char> = right_perm.iter().map(|&i| ib[i]).collect();

    let (med_gemm, iqr_gemm) = bench(|| {
        let mut c_arr = StridedArray::<f64>::col_major(&c_dims);
        strided_einsum2::einsum2_into_owned(
            c_arr.view_mut(), a_contig.clone(), b_contig.clone(),
            &ic, &ia_canon, &ib_canon, 1.0, 0.0, false, false,
        ).unwrap();
    }, warmup, nruns);
    println!("einsum2 (contiguous, ~GEMM):   {:.3} ms (IQR {:.3} ms)", med_gemm, iqr_gemm);

    // --- Summary ---
    println!("\n--- Summary ---");
    println!("Full einsum2 (scattered):  {:.3} ms", med);
    println!("  copy B (dominant):       {:.3} ms ({:.0}%)", med_copy_b, med_copy_b / med * 100.0);
    println!("  copy A:                  {:.3} ms", med_copy_a);
    println!("  GEMM only (~):           {:.3} ms ({:.0}%)", med_gemm, med_gemm / med * 100.0);
}
